newton/examples/diffsim/example_diffsim_walker.py

```diff
@@ -0,0 +1,330 @@
+        if wp.get_device().is_cuda:
+            with wp.ScopedCapture() as capture:
+                self.forward_backward()
+            self.graph = capture.graph
+        else:
+            self.graph = None
+
+    def forward_backward(self):
+        self.tape = wp.Tape()
+        with self.tape:
+            for i in range(self.sim_steps):
+                self.forward(i)
+        self.tape.backward(self.loss)
+
+    def forward(self, frame):
+        # build sinusoidal input phases
+        wp.launch(kernel=compute_phases, dim=self.phase_count, inputs=[self.phases[frame], self.sim_time])
+
+        # apply linear network with tanh activation
+        wp.launch_tiled(
+            kernel=network,
+            dim=math.ceil(self.model.tet_count / TILE_TETS),
+            inputs=[self.phases[frame].reshape((self.phase_count, 1)), self.weights],
+            outputs=[self.tet_activations[frame].reshape((self.model.tet_count, 1))],
+            block_dim=TILE_THREADS,
+        )
+        self.control.tet_activations = self.tet_activations[frame]
+
+        # run simulation loop
+        for i in range(self.sim_substeps):
+            t = frame * self.sim_substeps + i
+            self.states[t].clear_forces()
+            self.solver.step(
+                self.states[t],
+                self.states[t + 1],
+                self.control,
+                self.contacts,
+                self.sim_dt,
+            )
+            self.sim_time += self.sim_dt
+
+        # compute center of mass velocity
+        wp.launch(
+            com_kernel,
+            dim=self.model.particle_count,
+            inputs=[
+                self.states[(frame + 1) * self.sim_substeps].particle_qd,
+                self.model.particle_count,
+                self.coms[frame],
+            ],
+            outputs=[],
+        )
+
+        # compute loss
+        wp.launch(loss_kernel, dim=1, inputs=[self.coms[frame], self.loss], outputs=[])
+
+    def step(self):
+        if self.graph is not None:
+            wp.capture_launch(self.graph)
+        else:
+            self.forward_backward()
+
+        # optimization
+        self.optimizer.step([self.weights.grad.flatten()])
+
+        if self.verbose:
+            loss = self.loss.numpy()
+            print(f"Iteration {self.train_iter}: {loss}")
+
+        # reset sim
+        self.sim_time = 0.0
+        self.states[0] = self.model.state(requires_grad=True)
+
+        # clear grads and zero arrays for next iteration
+        self.tape.zero()
+        self.loss.zero_()
+        for i in range(self.sim_steps):
+            self.coms[i].zero_()
+
+        self.train_iter += 1
+
+    def render(self):
+        # draw bear training run
+        for i in range(self.sim_steps + 1):
+            state = self.states[i * self.sim_substeps]
+
+            self.viewer.begin_frame(self.frame * self.frame_dt)
+            self.viewer.log_state(state)
+            self.viewer.end_frame()
+
+            self.frame += 1
+
```

Comment:
Now that https://github.com/newton-physics/newton/pull/836 has been merged, can you add a useful `test()` method here similar to the other diffsim/particle examples?